### DISCUSSION TITLE
Lima and Extending cloud-init Templates ?

### DISCUSSION THREAD

**Discussion by sendit2me** [lima-vm/lima#1520] [Q&A] [answered]
I am trying to inject into the lima "start" the ability to add additional information to the cloud-init configuration (cidata.iso) the only documentation I can see is this is only available internally ? (docs/internal.md)

is this correct ?

I have a similar style setup in multipass, but want to adapt it to lima, so there is some cross processor architecture functionality. finding the arm64-linux quite problematic .. 

Also given qemu 8.0 is just release, was wondering if the cross-processor emulation is still really bad, I am mainly only looking at headless operations .

thanks

any input / direction would be appreciated.

**Comment by afbjorklund** [comment]
> if the cross-processor emulation is still really bad

when you say "bad", do you mean slow ? it should be ok

----

Some examples would make it easier to guess what you are trying to do, or if it would be a feature that would be nice to have

**Reply by sendit2me** [reply]
> > if the cross-processor emulation is still really bad
> 
> when you say "bad", do you mean slow ? it should be ok

Sorry, "bad" is just based on all the benchmarks I can see online, they are mainly 2022 (early)
* I am in the process of running a simple "Phoronix Test Suite" mainly the cpu one, to see roughly

I want to setup up a pc-agnostic dev env, multiple types, specified by cloud-init which is yaml, so lend itself to being "smashed" together (yaml being composable). but am early days on the process with lima. Currently only investigations, the nice work done in lima with regards to formalised integrations for network/shares etc, and pure qemu based bash scripts.

*currently I have a set of cloud-init files (or sections of) which of composed together based on a command-line (think -f python -f vscode) which is then bound to the booting vm, and then subsequently installed. if something like this can be done with lima, so that the resulting VM is always the same processor arch, would be nice. and the ability to run this on any host (arch, so either amd64 or m1) without the pain.

*avoiding, runs on my other machine :)

**Reply by afbjorklund** [reply]
It would be better with multi-arch images, so that you can run them in a native architecture VM ?

But as long you expect the 10x overhead (emu), instead of the usual 5% (virt), it should not be "bad"

**Reply by sendit2me** [reply]
I have found/had a lot of issues running some "similar" software which has compatibility/not-supported issues on arm64/linux.

On a M1 running multipass with ubuntu, I get quite a few issues, vs running the same setup on a  x86_64 intel. (multipass cloud-init'ed)
I think a lot comes down to (I think) the support for arm64 on Darwin, vs the arm64 and linux.

I was thinking of some way to take the cloud-init that is used in lima, with the ability to incorporate another cloud-init. I current yaml merge multiple documents, for different layers of the dev environment I am trying to compose. so would look to taking the final layer, and merging it with the layer from lima. which can then be run on the VM.
I see lima has script options, so I may see about (I'm not a go developer) creating a yaml attribute, something which would point to an "existing cloud-init" user/meta data or file reference and combined get turned into a cidata.iso for use in the base image

**Reply by sendit2me** [reply, marked as answer]
Hi

So I did a bit of tinkering (go is not my main language) 
I forked the repo and have since done one extra commit,
https://github.com/sendit2me/lima 

It basically add a map of strings [filename: ref] which allows you to reference a file "user-data" "meta-data" etc which will be yaml merged with the matching file (if exists) in the cidata iso, 
this allows for more complex cloud-init templates, which comes in quite handy
